src/Plutarch/Extra/Ord.hs

```diff
@@ -0,0 +1,804 @@
+                _ -> pcons # x #$ self # x' # xs' # y # ys
+
+{- | Sort a list-like by the 'POrd' instance of its contents.
+
+ This uses a [merge sort](https://en.wikipedia.org/wiki/Merge_sort)
+ implementation, which is also
+ [stable](https://en.wikipedia.org/wiki/Sorting_algorithm#Stability). As this
+ is a comparison sort, it requires a linearithmic ($n \log(n)$) number of
+ comparisons to complete.
+
+ @since 3.6.0
+-}
+psort ::
+    forall (a :: S -> Type) (ell :: (S -> Type) -> S -> Type) (s :: S).
+    (POrd a, PElemConstraint ell a, PListLike ell) =>
+    Term s (ell a :--> ell a)
+psort = phoistAcyclic $ plam (psortBy # pfromOrd @a #)
+
+{- | As 'psort', but uses a custom 'PComparator' for order comparisons.
+
+ @since 3.6.0
+-}
+psortBy ::
+    forall (a :: S -> Type) (ell :: (S -> Type) -> S -> Type) (s :: S).
+    (PElemConstraint ell a, PListLike ell) =>
+    Term s (PComparator a :--> ell a :--> ell a)
+psortBy = phoistAcyclic $
+    plam $ \cmp xs ->
+        pmergeAllUnsafe
+            cmp
+            (const (pconvertLists #))
+            pmergeUnsafe
+            #$ pmergesort
+            # cmp
+            # xs
+
+{- | As 'psort', but also throws out all duplicates according to the 'PEq'
+ instance of its argument's contents.
+
+ @since 3.6.0
+-}
+pnubSort ::
+    forall (a :: S -> Type) (ell :: (S -> Type) -> S -> Type) (s :: S).
+    (POrd a, PElemConstraint ell a, PListLike ell) =>
+    Term s (ell a :--> ell a)
+pnubSort = phoistAcyclic $ plam (pnubSortBy # pfromOrd @a #)
+
+{- | As 'pnubSort', but uses a custom 'PComparator' for both ordering and
+ equality.
+
+ @since 3.6.0
+-}
+pnubSortBy ::
+    forall (a :: S -> Type) (ell :: (S -> Type) -> S -> Type) (s :: S).
+    (PElemConstraint ell a, PListLike ell) =>
+    Term s (PComparator a :--> ell a :--> ell a)
+pnubSortBy = phoistAcyclic $
+    plam $ \cmp xs ->
+        pmergeAllUnsafe
+            cmp
+            (\cmp' xs' -> pconvertLists #$ pnubUnsafe # cmp' # xs')
+            pmergeUnsafeNoDupes
+            #$ pmergesort
+            # cmp
+            # xs
+
+-- Helpers
+
+-- Merges nested PLists and repacks into a list-like of your choice, given a
+-- method of merging the PLists themselves.
+--
+-- Only call this if you are _sure_ that the 'inner' PLists are sorted according
+-- to the argument comparator!
+pmergeAllUnsafe ::
+    forall (a :: S -> Type) (ell :: (S -> Type) -> S -> Type) (s :: S).
+    (PElemConstraint ell a, PListLike ell) =>
+    Term s (PComparator a) ->
+    (forall (s' :: S). Term s' (PComparator a) -> Term s' (PList a) -> Term s' (ell a)) ->
+    (forall (s' :: S). Term s' (PComparator a) -> Term s' (PList a) -> Term s' (PList a) -> Term s' (PList a)) ->
+    Term s (PList (PList a) :--> ell a)
+pmergeAllUnsafe cmp whenSingleton whenMerging = pfix #$ plam $ \self xs ->
+    phandleList xs pnil $ \x' xs' ->
+        phandleList xs' (whenSingleton cmp x') $ \_ _ ->
```

Comment:
`whenSingleton` is pointless, same efficiency as doing `pconvertLists` after the fact. Should produce the output list type when doing the last merge instead. You could probably even do this with the same `whenMerging` if it's polymorphic in its output list type, it just needs to be independent of its input list type.